README.md
<h1 align="center">
    quadtree-lib<br>
	<br>
	<a href="http://elbywan.github.io/quadtree-lib/demo/collisions/collisions.html" target="_blank">
		<img alt="quatree gif" src="http://elbywan.github.io/quadtree-lib/assets/quadtree.gif" width="120px"/>
	</a><br>
	<a href="https://travis-ci.org/elbywan/quadtree-lib" target="_blank" style="margin-right: 10px">
		<img alt="Build Status" src="https://travis-ci.org/elbywan/quadtree-lib.svg?branch=master"/>
	</a>
	<a href="https://coveralls.io/github/elbywan/quadtree-lib?branch=master" target="_blank" style="margin-right: 10px">
		<img alt="Coverage Status" src="https://coveralls.io/repos/github/elbywan/quadtree-lib/badge.svg?branch=master"/>
	</a>
	<a href="https://www.npmjs.com/package/quadtree-lib" target="_blank">
		<img alt="npm version" src="https://badge.fury.io/js/quadtree-lib.svg"/>
	</a>
</h1>
<h4 align="center">
	Quadtree-lib is an easy to use, developer friendly quadtree library which
	contains many helper methods to add, remove, iterate, filter, simulate
	collisions over 2d elements and more.
</h4>

If you are already familiar with quadtrees, then you should perfectly understand
how to use this library.

Otherwise, there are many online articles
([wikipedia does the job](https://en.wikipedia.org/wiki/Quadtree)) which explain
the advantages of using the quadtree datastructure in certain situations.

If you want to see the library in action :
- [Collision demo](http://elbywan.github.io/quadtree-lib/demo/collisions/collisions.html)
- [Movement demo](http://elbywan.github.io/quadtree-lib/demo/movement/movement.html)
- [Mass demo](http://elbywan.github.io/quadtree-lib/demo/basic/basic.html)

## Setup

### Using npm / yarn

From the command line :

`npm install quadtree-lib` or `yarn add quadtree-lib`

### Using bower

`bower install quadtree-lib`

### Using gulp

In your favorite terminal :

```bash
# 1°clone the repo
git clone https://github.com/elbywan/quadtree-lib
# 2° change dir
cd quadtree-lib
# 3° build the library
gulp
# 4° build the documentation
gulp doc
# 5° run performance tests
gulp perf
# 6° profit
```

## Usage

### Import

**This library is bundled in UMD format.**

Examples :

- Import using commonjs :
```javascript
Quadtree = require("quadtree-lib")
```

- Import globally with a script tag :
```html
<script src="path/to/quadtree-lib.min.js"></script>
```

### Init

First step is to initialize a new Quadtree object.

```javascript
var quadtree = new Quadtree({
    width: 500,
    height: 500,
    maxElements: 5 //Optional
})
```

`width` and `height` are mandatory attributes.

`maxElements` (default 1) is the maximum number of elements contained in a leaf before it
splits into child trees.

### For typescript users

A set of declaration files (.d.ts) is included, which means that you have access to auto-completion and embedded documentation in your favorite IDE.

If you are using the library **globally** with a `<script>` tag, add the following declaration import :
```typescript
/// <reference types="quadtree-lib" />
```

Otherwise, if you are using the **commonjs** way :
```typescript
import * as Quadtree from "quadtree-lib"
```

### Adding elements

Elements must be objects, with coordinates set.

Optionally, you can pass a boolean argument which, if set to `true`, will
remove/push the object into the quadtree each time its coordinates or dimensions
are set *(ex: item.x = ... or item.width = ...)*.

*Without this flag, x / y / width / height properties should* **not** *be
changed  after insertion.*

```javascript
quadtree.push({
    x: 10,      //Mandatory
    y: 10,      //Mandatory
    width: 1,   //Optional, defaults to 1
    height: 2   //Optional, defaults to 1
}, true) //Optional, defaults to false
```

To insert an array of elements, use the **pushAll** method which is faster than inserting each element with push.

```javascript
quadtree.pushAll([
    {x: 1, y: 1},
    {x: 2, y: 2}
    // ... //
])
```

### Removing elements

Removes an item by reference.

```javascript
quadtree.remove(item)
```

### Clearing the tree

Removes the tree contents and restores it to pristine state.

```javascript
quatree.clear()
```

### Filtering the tree

Filters the quadtree and returns a **clone** containing only the elements
determined by a predicate function.

```javascript
var filtered = quadtree.filter(function(element){
    return element.x > 50
})
```

*Opposite: quadtree.reject*

### Retrieve colliding elements

Gets every element that collides with the parameter 2d object.

```javascript
var colliding = quadtree.colliding({
    x: 10,
    y: 10,
    width: 5, //Optional
    height: 5 //Optional
})
```

The default collision function is a basic bounding box algorithm.
You can change it by providing a function as a second argument.

```javascript
var colliding = quadtree.colliding({
    x: 10,
    y: 10
}, function(element1, element2){
    return // Place collision algorithm here //
})

```

### Perform an action on colliding elements

Performs an action on every element that collides with the parameter 2d object.

```javascript
onCollision({
    x: 10,
    y: 20
}, function(item) {
    /* Action on colliding item */

    // As with all iterative methods, modifying the quadtree or its contents is discouraged. //
}, function(element1, element2){
    /* Optional custom collision algorithm */
    return // Place predicate here //
})
```

### Retrieve by properties

Gets every element that match the parameter properties.

```javascript
quadtree.push({x: 0, y: 0, animal: 'rabbit'})
var match = quadtree.where({
    animal: 'rabbit'
})
```

*Alias : quadtree.get*

### Retrieve by predicate

Gets every element that validate the given predicate.

```javascript
quadtree.find(function(element){
    return element.color === 'red' //Example
})
```

### Iterate over the elements

Performs an action on each element of the Quadtree (breadth first traversal).

```javascript
quadtree.each(function(element){
    console.log(element.color)

    // As with all iterative methods, modifying the quadtree or its contents is discouraged. //
})
```

*Like some other data structures, it is strongly discouraged to update Quadtree elements (especially coordinates / dimensions) or the Quadtree itself while iterating on it.*

### Visit the tree nodes

Visits each node of the quadtree. (meaning subtrees)

```javascript
quadtree.visit(function(){
    // This function is called once for each node.
    // *this* is a pointer to the current node.
    console.log(this.contents)

    // As with all iterative methods, modifying the quadtree or its contents is discouraged. //
})
```

### Pretty print

Outputs the tree and its contents in an eye friendly format.

```javascript
var quadtree = new Quadtree({
  width: 10,
  height: 10,
  maxElements: 1
});

var elementArray = [
  element0 = {
    x: 0,
    y: 0,
    toString: function() {
      return 0;
    }
  }, element1 = {
    x: 3,
    y: 3,
    toString: function() {
      return 1;
    }
  }
];

quadtree.pushAll(elementArray);

console.log(quadtree.pretty());
```

Console output :

```
| ROOT
| ------------
└──┐
   | NW
   | ------------
   └──┐
      | SE
      | ------------
      | * Leaf content *
      |   1
      | NW
      | ------------
      | * Leaf content *
      |   0
```


## Further documentation

You can find the annotated source code [here](http://elbywan.github.io/quadtree-lib/).

*Generated with Docco.*

## License

#### The MIT License (MIT)

Copyright (c) 2015 Julien Elbaz

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# Change log - quadtree-lib

## [1.0.9](https://github.com/elbywan/quadtree-lib/compare/1.0.8...1.0.9)

##### Fixed

- [Relaxed the QuadtreeItem type](https://github.com/elbywan/quadtree-lib/commit/5da28915ccd1d5412bb30e1c7baf199400d9ffe0)
- [Fix mass demo input reference](https://github.com/elbywan/quadtree-lib/commit/6551d58d0b37c2ae3c7a200cc977a1c8e2d10197)

## [1.0.8](https://github.com/elbywan/quadtree-lib/compare/1.0.7...1.0.8)

##### Added

- [Clear function](https://github.com/elbywan/quadtree-lib/commit/bff7a534185c3e4b127d7c5b9ad3d712f35e2e8c)
- [Javascript linter (Eslint)](https://github.com/elbywan/quadtree-lib/commit/c6ec2412926a218880e6bc8d3877a6d0fcc5a1cf)
- [Movement demo](https://github.com/elbywan/quadtree-lib/commit/a55695507a02597bcca059f35e354c5c87eb8d06)
- [Demo .js refactoring](https://github.com/elbywan/quadtree-lib/commit/2ef03608849ee0cc8cf09cad8fbe82a1b1ad3aac)

## [1.0.7](https://github.com/elbywan/quadtree-lib/compare/1.0.6...1.0.7)

##### Added

- [Coffeescript linter (http://www.coffeelint.org/)](https://github.com/elbywan/quadtree-lib/commit/40e774335d22b0b73a3c5017aa374e10a46f1a0f)
- [Unobserve elements on removal](https://github.com/elbywan/quadtree-lib/commit/f6ecc001c66891ea1f3713b1afcd3e3e19b33be4)
- [Add onCollision method](https://github.com/elbywan/quadtree-lib/commit/ec4f92cb77d3f05dfe9b45bd16c676f18b309559)

## [1.0.6](https://github.com/elbywan/quadtree-lib/compare/1.0.5...1.0.6)

##### Added

- [Add changelog](https://github.com/elbywan/quadtree-lib/commit/a126c8e38f42174237b6ebaa5c7b85e42ed1706a)
- [[Feature] Allow negative quadtree coordinates](https://github.com/elbywan/quadtree-lib/commit/18fcec76d68c1b05c55672b6767682ce4103ded6)

## [1.0.5](https://github.com/elbywan/quadtree-lib/compare/1.0.4...1.0.5)

##### Added

- [[Feature] #2 : Allow negative x and y values](https://github.com/elbywan/quadtree-lib/commit/5cf0e140d66fde00d5d61810b26cf3b018b45526)

##### Changed

- [upgrade devDeps](https://github.com/elbywan/quadtree-lib/commit/dd7fd1e62e092e7d6fba7953c827630eb961780e)

##### Fixed

- [readme typos](https://github.com/elbywan/quadtree-lib/commit/5b7b7fe6ac8838ad831c864d4242cc2eefb7ff37)

## [1.0.4](https://github.com/elbywan/quadtree-lib/compare/1.0.3...1.0.4)

##### Fixed

- [fix critical bundling issue](https://github.com/elbywan/quadtree-lib/commit/4135ab76872e3d0ec43c0f52f66fdeac56bc3802)

## [1.0.3](https://github.com/elbywan/quadtree-lib/compare/1.0.2...1.0.3)

##### Added

- [proper UMD bundling](https://github.com/elbywan/quadtree-lib/commit/d8a775f7c5ab8348d08e4728f21213a373ba383c)

##### Fixed

- [fixed sourcemaps ](https://github.com/elbywan/quadtree-lib/commit/a966df72026376c51621b6116b7862baebb2e9bb)
- [fixed assets](https://github.com/elbywan/quadtree-lib/commit/5debe73d1613ab8e62761ce1be44eeae6fd4e4f4)

## [1.0.2](https://github.com/elbywan/quadtree-lib/compare/1.0.1...1.0.2)

##### Added

- [Typescript support with a declaration file](https://github.com/elbywan/quadtree-lib/commit/1064cfda571f526276aab5b3eb602d53eee14692)
- [[doc] Add github corner ](https://github.com/elbywan/quadtree-lib/commit/dc8a3eb67323a277aef86371add56a1026278c9d)
- [[demo] collisions / add counters ](https://github.com/elbywan/quadtree-lib/commit/3a9fb0ef069f0154214b3567fdbc240c80b0895b)

##### Fixed

- [Collision demo : prevent div resize out of bounds.](https://github.com/elbywan/quadtree-lib/commit/ca736547e145c138129113153a0b4f427d8f2f31)
- [[demo] collisions / add buttons & chrome bounding box fix ](https://github.com/elbywan/quadtree-lib/commit/32ea35000efd6059c45924a22b048fd8b1cae771)

## [1.0.1](https://github.com/elbywan/quadtree-lib/compare/1.0.0...1.0.1)

##### Added

- [Pretty print function.](https://github.com/elbywan/quadtree-lib/commit/d4f9c13fc32679ecfeafdd13ee06e9e39e728258)

##### Fixed

- [fix removal not splicing the right array ](https://github.com/elbywan/quadtree-lib/commit/4a80caed3e2d93c8164d111b1d0f417911d911cc)

## 1.0.0

#### First stable version.


demo/common.js
/* global window, document */

window.randomNb = function(min, max) {
    if(min >= max)
        throw new Error("min must be < max")
    return Math.floor(Math.random() * (max - min)) + min
}

window.randomColor = function() {
    var color = ((1 << 24) * Math.random() | 0).toString(16)
    var pad = function(str, length) {
        if(str.length < length) {
            str = "0" + str
            return pad(str, length)
        } else {
            return str
        }
    }
    return "#" + pad(color, 6)
}

window.drawSquare = function(elt, fill, context) {
    context.beginPath()
    context.moveTo(elt.x, elt.y)
    context.lineTo(elt.x + (elt.width ? elt.width : 1), elt.y)
    context.lineTo(elt.x + (elt.width ? elt.width : 1), elt.y + (elt.height ? elt.height : 1))
    context.lineTo(elt.x, elt.y + (elt.height ? elt.height : 1))
    context.closePath()
    context.stroke()
    if(fill)
        context.fill()
}

window.drawQuadtree = function(tree, fill, context) {
    var halfWidth  = Math.max(Math.floor(tree.width  / 2), 1)
    var halfHeight = Math.max(Math.floor(tree.height / 2), 1)

    window.drawSquare({
        x: tree.x,
        y: tree.y,
        width:  halfWidth,
        height: halfHeight
    }, fill, context)
    window.drawSquare({
        x: tree.x + halfWidth,
        y: tree.y,
        width:  halfWidth,
        height: halfHeight
    }, fill, context)
    window.drawSquare({
        x: tree.x,
        y: tree.y + halfHeight,
        width:  halfWidth,
        height: halfHeight
    }, fill, context)
    window.drawSquare({
        x: tree.x + halfWidth,
        y: tree.y + halfHeight,
        width:  halfWidth,
        height: halfHeight
    }, fill, context)
}

window.makeMovable = function(element, boundaryElement, callbacks) {
    var extractPos = function(event) {
        return {
            x: event.clientX ? event.clientX : event.changedTouches[0].clientX,
            y: event.clientY ? event.clientY : event.changedTouches[0].clientY
        }
    }

    var resizeAction = function(event) {
        var targetRect = element.getBoundingClientRect()
        var position = extractPos(event)

        if(boundaryElement) {
            var boundaries = boundaryElement.getBoundingClientRect()
            element.style.height = Math.max(5, Math.min(position.y - targetRect.top,  boundaries.bottom - targetRect.top))
            element.style.width  = Math.max(5, Math.min(position.x - targetRect.left, boundaries.right  - targetRect.left))
        } else {
            element.style.height = Math.max(5, position.y - targetRect.top)
            element.style.width  = Math.max(5, position.x - targetRect.left)
        }

        if(callbacks && callbacks.onResize && typeof callbacks.onResize === "function") {
            callbacks.onResize()
        }
    }
    var lastMovePos = null
    var clickAction = function(event) {
        var targetRect = element.getBoundingClientRect()
        var position = extractPos(event)

        if(Math.abs(position.y - targetRect.bottom) < 12 && Math.abs(position.x - targetRect.right < 12)) {
            document.getElementsByTagName("body")[0].addEventListener("mousemove", resizeAction)
            document.getElementsByTagName("body")[0].addEventListener("touchmove", resizeAction)
        } else {
            document.getElementsByTagName("body")[0].addEventListener("mousemove", moveAction)
            document.getElementsByTagName("body")[0].addEventListener("touchmove", moveAction)
            lastMovePos = { x: position.x, y: position.y }
        }
        event.preventDefault()
    }
    var moveAction = function(event) {
        var position = extractPos(event)

        if(boundaryElement && boundaryElement.contains(event.target)) {
            var boundaries = boundaryElement.getBoundingClientRect()

            element.style.top  = parseInt(window.getComputedStyle(element).top)  + position.y - lastMovePos.y
            element.style.left = parseInt(window.getComputedStyle(element).left) + position.x - lastMovePos.x
            element.style.top  = Math.max(0, Math.min(parseInt(element.style.top),  boundaries.height - parseInt(window.getComputedStyle(element).height)))
            element.style.left = Math.max(0, Math.min(parseInt(element.style.left), boundaries.width  - parseInt(window.getComputedStyle(element).width)))

            lastMovePos = { x: position.x, y: position.y }

            if(callbacks && callbacks.onMove && typeof callbacks.onMove === "function") {
                callbacks.onMove()
            }
        } else if(!boundaryElement) {
            element.style.top  = parseInt(window.getComputedStyle(element).top)  + position.y - lastMovePos.y
            element.style.left = parseInt(window.getComputedStyle(element).left) + position.x - lastMovePos.x

            lastMovePos = { x: position.x, y: position.y }

            if(callbacks && callbacks.onMove && typeof callbacks.onMove === "function") {
                callbacks.onMove()
            }
        }
    }

    element.addEventListener("mousedown",  clickAction)
    element.addEventListener("touchstart", clickAction)

    document.getElementsByTagName("body")[0].addEventListener("mouseup", function() {
        document.getElementsByTagName("body")[0].removeEventListener("mousemove", resizeAction)
        document.getElementsByTagName("body")[0].removeEventListener("mousemove", moveAction)
    })
    document.getElementsByTagName("body")[0].addEventListener("touchend", function() {
        document.getElementsByTagName("body")[0].removeEventListener("touchmove", resizeAction)
        document.getElementsByTagName("body")[0].removeEventListener("touchmove", moveAction)
    })

}


demo/basic/basic.js
var canvas = document.getElementById("quadtree_canvas")
var layer = document.getElementById("layer_canvas")
var container = document.getElementById("canvas-container")
var counter = document.getElementById("counter")
var ctx = canvas.getContext("2d")
var layerCtx = layer.getContext("2d")
var width = Math.min(container.clientWidth, window.innerWidth)
var height = width === window.innerWidth ? window.innerWidth : container.clientHeight
DELAY = 1
canvas.width = width
canvas.height = height
layer.width = width
layer.height = height

quadtreeColor = 'rgba(120, 144, 156, 0.1)'
eltColor = 'rgba(229, 57, 53, 1)'
oversizeColor = 'rgba(136, 14, 79, 1)'
collidingColor = '#F57F17'

eltSizeQuota = 25
eltDrawMult = 100
eltIncrement = 10

document.addEventListener('DOMContentLoaded', function () {
    init()
    updateCanvas()
    updateLayer()
})

var init = function(){
    canvas.width = width
    canvas.height = height
    container.style.height = height+"px"
    ctx.clearRect(0, 0, width, height)
    ctx.lineWidth = 1
    layer.width = width
    layer.height = height
    layerCtx.clearRect(0, 0, width, height)
    layerCtx.lineWidth = 1
    quadtree = new Quadtree({
        width: width,
        height: height
    })
    updateCounter()
}

var updateCounter = function(){
    counter.innerHTML = quadtree.size + " elements"
}

var randomizeElement = function(){
    var squareSize = randomNb(0, Math.min(width, height) / eltSizeQuota)
    return {
        x: randomNb(0, width),
        y: randomNb(0, height),
        width: squareSize,
        height: squareSize,
        color: eltColor//"#"+((1<<24)*Math.random()|0).toString(16)
    }
}

var updateCanvas = function(){
    ctx.clearRect(0, 0, width, height)
    quadtree.visit(function(){
        ctx.strokeStyle = quadtreeColor
        drawQuadtree(this)
        ctx.fillStyle = eltColor
        for(i in this.contents)
            drawSquare(this.contents[i], true)
        ctx.fillStyle = oversizeColor
        for(i in this.oversized)
            drawSquare(this.oversized[i], true)
    })
}

var updateLayer = function(){
    layerCtx.clearRect(0, 0, width, height)
    if(window.mousePos){
        layerCtx.strokeStyle = collidingColor
        layerCtx.fillStyle = collidingColor
        layerCtx.beginPath()
        layerCtx.moveTo(0, mousePos.y)
        layerCtx.lineTo(width, mousePos.y)
        layerCtx.closePath()
        layerCtx.stroke()
        layerCtx.beginPath()
        layerCtx.moveTo(mousePos.x, 0)
        layerCtx.lineTo(mousePos.x, height)
        layerCtx.closePath()
        layerCtx.stroke()
        quadtree.colliding(mousePos).forEach(function(elt){
            drawSquare(elt, true, layerCtx)
        })
    }
}

var addElements = function(){
    var elementArray = []
    for(var i = 0; i < eltIncrement; i++)
        elementArray.push(randomizeElement())
    quadtree.pushAll(elementArray)
    updateCanvas()
    updateLayer()
    updateCounter()
}

var unregisterMouse = function(){
    delete window.mousePos
    updateLayer()
}

var hoverMouse = function(event){
    mousePos = {
        x: event.offsetX || (event.offsetX === 0 ? 0 : event.changedTouches[0].clientX - event.target.getBoundingClientRect().left),
        y: event.offsetY || (event.offsetY === 0 ? 0 : event.changedTouches[0].clientY - event.target.getBoundingClientRect().top)
    }
    if(mousePos.x < 0 || mousePos.y < 0)
        return
    event.stopPropagation()
    event.preventDefault()
    updateLayer()
}


typings/quadtree.d.ts
export as namespace Quadtree
export = Quadtree

declare class Quadtree<T extends Quadtree.QuadtreeItem> {

    /**
     * The Quadtree class constructor.
     * @param options Object containing the tree properties.
     *  - width, height : dimensions.
     *  - maxElements: maximum number of elements per leaf, further elements cause a split.
     *  - x, y : these are used internally by the library to position subtrees.
     */
    constructor(options: {
        x?: number,
        y?: number,
        width: number,
        height: number,
        maxElements?: number
    })

    /**
     * Removes all elements from the quadtree and restores pristine state.
     */
    public clear() : void

    /**
     * Add an element to the quadtree.
     *
     * Elements can be observed to reorganize them into the quadtree automatically whenever their coordinates or dimensions are set (for ex. obj.x = ...).
     *
     * @param item Element to add.
     * @param doObserve Observes the element and removes/add it to the quadtree when its dimensions/coordinates are modified.
     */
    public push(item: T, doObserve?: boolean): Quadtree<T>
    /**
     * Push an array of elements.
     *
     * @param items Array of elements.
     * @param doObserve Observes the elements and removes/add them to the quadtree when their dimensions/coordinates are modified.
     */
    public pushAll(items: Array<T>, doObserve?: boolean) : Quadtree<T>
    /**
     * Removes an element from the quadtree.
     */
    public remove(item: T) : boolean
    /**
     * Returns an array of elements which collides with the `item` argument.
     *`item` being an object having x, y, width & height properties.
     *
     *The default collision function is a basic bounding box algorithm.
     *You can change it by providing a function as a second argument.
     *```javascript
     *colliding({x: 10, y: 20}, function(element1, element2){
     *    return //Predicate
     *})
     *```
     */
    public colliding(item: Quadtree.QuadtreeItem, collisionFunction?: (elt1: T, elt2: T) => boolean) : Array<T>
    /** Performs an action on elements which collides with the `item` argument.
    * `item` being an object having x, y, width & height properties.
    *
    * The default collision function is a basic bounding box algorithm.
    * You can change it by providing a function as a third argument.
    *```javascript
    *onCollision(
    *   { x: 10, y: 20 },
    *   function(item) {
    *       // stuff //
    *   },
    *   function(element1, element2) {
    *       return // Place predicate here //
    *})
    * ```
    */
    public onCollision(item: Quadtree.QuadtreeItem, callback: Function, collisionFunction ?: (elt1: T, elt2: T) => boolean) : void
    /**
     * Returns an array of elements that match the `query` argument.
     */
    public get(query: Object): Array<T>
    /**
     * Returns an array of elements that match the `query` argument.
     */
    public where(query: Object) : Array<T>
    /**
     * For each element of the quadtree, performs the `action` function.
     *
     * ```javascript
     * quad.each(function(item){ console.log(item) })
     * ```
     */
    public each(action: (elt: T) => void) : Quadtree<T>
    /**
     * Returns an array of elements which validate the predicate.
     */
    public find(predicate: (elt: T) => boolean) : Array<T>
    /**
     * Returns a **cloned** `Quadtree` object which contains only the elements that validate the predicate.
     */
    public filter(predicate: (elt: T) => boolean) : Quadtree<T>
    /**
     * Opposite of filter.
     */
    public reject(predicate: (elt: T) => boolean) : Quadtree<T>
    /**
     * Visits each tree & subtree contained in the `Quadtree` object.
     * For each node, performs the `action` function, inside which `this` is bound to the node tree object.
     */
    public visit(action: (tree: Quadtree<T>) => void) : Quadtree<T>
    /**
     * Pretty printing function.
     */
    public pretty() : string
}

declare namespace Quadtree {
    export type QuadtreeItem = { x: number, y: number, width?: number, height?: number, [extras: string]: any }
}


demo/movement/movement.js
/* global window, document */

window.demo = (function() {

    var canvas = document.getElementById("quadtree_canvas")
    var ctx = canvas.getContext("2d")
    var container = document.getElementById("canvas-container")
    var counter = document.getElementById("counter")
    var width = Math.min(container.clientWidth, window.innerWidth)
    var height = width === window.innerWidth ? window.innerWidth : container.clientHeight

    var movementDemo = {

        quadtreeColor: "rgba(120, 144, 156, 0.1)",
        eltColor: "rgba(136, 14, 79, 1)",
        eltNb: 50,
        elts: [],
        quadtree: null,

        init: function() {
            canvas.width = width
            canvas.height = height
            container.style.height = height + "px"
            ctx.clearRect(0, 0, width, height)
            ctx.lineWidth = 1
            this.quadtree = new window.Quadtree({
                width: width,
                height: height
            })
            this.elts = []
            for(var i = 0; i < this.eltNb; i++) {
                this.elts.push(this.randomElement())
            }
            this.quadtree.pushAll(this.elts)
            this.updateCanvas()
            this.updateCounter()
        },

        randomElement: function() {
            var squareWidth = window.randomNb(5, 20)

            return {
                x: window.randomNb(0, width - squareWidth),
                y: window.randomNb(0, height - squareWidth),
                width: squareWidth,
                height: squareWidth,
                speed: window.randomNb(1, 5),
                direction: [ window.randomNb(-1, 2) || -1, window.randomNb(-1, 2) || 1 ],
                delay: window.randomNb(1, 3),
                delayIncrement: 0,
                color: this.eltColor
            }
        },

        addElements: function() {
            var elementArray = []
            for(var i = 0; i < this.eltNb; i++) {
                var elt = this.randomElement()
                elementArray.push(elt)
                this.elts.push(elt)
            }
            this.quadtree.pushAll(elementArray)
            this.updateCounter()
        },

        updateCanvas: function() {
            ctx.clearRect(0, 0, width, height)
            var that = this
            this.quadtree.visit(function() {
                ctx.strokeStyle = that.quadtreeColor
                window.drawQuadtree(this, false, ctx)
                for(var i in this.contents) {
                    ctx.fillStyle = this.contents[i].color
                    window.drawSquare(this.contents[i], true, ctx)
                }
                for(var j in this.oversized) {
                    ctx.fillStyle = this.oversized[j].color
                    window.drawSquare(this.oversized[j], true, ctx)
                }
            })
        },

        updateCounter: function() {
            counter.innerHTML = this.quadtree.size + " elements"
        },

        movementTick: function() {
            var that = this
            this.elts.forEach(function(elt) {
                var collisions = that.quadtree.colliding(elt)
                if(elt.x < 0) {
                    if(elt.direction[0] < 0)
                        elt.direction[0] = -elt.direction[0]
                    elt.x = 0
                } else if(elt.x + elt.width > width) {
                    if(elt.direction[0] > 0)
                        elt.direction[0] = -elt.direction[0]
                    elt.x = width - elt.width
                } else if(elt.y < 0) {
                    if(elt.direction[1] < 0)
                        elt.direction[1] = -elt.direction[1]
                    elt.y = 0
                } else if(elt.y + elt.height > height) {
                    if(elt.direction[1] > 0)
                        elt.direction[1] = -elt.direction[1]
                    elt.y = height - elt.height
                } else if(collisions.length > 0) {
                    elt.color = window.randomColor()
                    var collision = collisions[0]
                    var edges = {}
                    if(collision.x <= elt.x && collision.x + collision.width  >= elt.x)  {
                        edges.left = collision.x + collision.width - elt.x
                    }
                    if(collision.x + collision.width >= elt.x + elt.width && collision.x <= elt.x + elt.width) {
                        edges.right = elt.x + elt.width - collision.x
                    }
                    if(collision.y <= elt.y && collision.y + collision.height  >= elt.y)  {
                        edges.top = collision.y + collision.height - elt.y
                    }
                    if(collision.y + collision.height >= elt.y + elt.height && collision.y <= elt.y + elt.height) {
                        edges.bottom = elt.y + elt.height - collision.y
                    }
                    var bestEdge = null
                    for(var key in edges) {
                        if(!bestEdge || bestEdge[1] > edges[key])
                            bestEdge = [ key, edges[key] ]
                    }
                    if(bestEdge) {
                        if(bestEdge[0] === "top") {
                            if(elt.direction[1] < 0)
                                elt.direction[1] = -elt.direction[1]
                            // elt.y = collision.y + collision.height
                        } else if(bestEdge[0] === "bottom") {
                            if(elt.direction[1] > 0)
                                elt.direction[1] = -elt.direction[1]
                            // elt.y = collision.y - elt.height
                        } else if(bestEdge[0] === "left") {
                            if(elt.direction[0] < 0)
                                elt.direction[0] = -elt.direction[0]
                            // elt.x = collision.x + collision.width
                        } else if(bestEdge[0] === "right") {
                            if(elt.direction[0] > 0)
                                elt.direction[0] = -elt.direction[0]
                            // elt.x = collision.x - elt.width
                        }
                    }
                }
            })
            var removedElts = []
            this.elts.forEach(function(elt) {
                elt.delayIncrement = (elt.delayIncrement + 1) % elt.delay
                if(elt.delayIncrement === 0) {
                    that.quadtree.remove(elt)
                    removedElts.push(elt)
                    elt.x += elt.direction[0] * elt.speed
                    elt.y += elt.direction[1] * elt.speed
                }
            })
            this.quadtree.pushAll(removedElts)
        }
    }

    document.addEventListener("DOMContentLoaded", function() {
        movementDemo.init()
    })

    var renderingLoop = function() {
        movementDemo.movementTick()
        movementDemo.updateCanvas()
        window.requestAnimationFrame(renderingLoop)
    }

    window.requestAnimationFrame(renderingLoop)

    return movementDemo
}())
